packages/edit-site/src/components/resizable-frame/index.js

```diff
@@ -0,0 +1,204 @@
+};
+
+function ResizableFrame( { isFull, children } ) {
+	const [ frameSize, setFrameSize ] = useState( {
+		width: '100%',
+		height: '100%',
+	} );
+	const [ isResizing, setIsResizing ] = useState( false );
+	const [ isHovering, setIsHovering ] = useState( false );
+	const [ isOversized, setIsOversized ] = useState( false );
+	const { setCanvasMode } = unlock( useDispatch( editSiteStore ) );
+	const initialAspectRatioRef = useRef( null );
+	const initialComputedWidthRef = useRef( null );
+	const FRAME_TRANSITION = { type: 'tween', duration: isResizing ? 0 : 0.5 };
+
+	// Calculate the frame size based on the window width as its resized.
+	const handleResize = ( event, direction, ref ) => {
+		const updatedWidth = ref.offsetWidth;
+		const updatedHeight = ref.offsetHeight;
+
+		if ( initialComputedWidthRef.current === null ) {
+			initialComputedWidthRef.current = updatedWidth;
+		}
+
+		// Set the initial aspect ratio if it hasn't been set yet.
+		if ( initialAspectRatioRef.current === null ) {
+			initialAspectRatioRef.current = updatedWidth / updatedHeight;
+		}
+
+		// Calculate the intermediate aspect ratio based on the current width.
+		const lerpFactor =
+			1 -
+			Math.max(
+				0,
+				Math.min(
+					1,
+					( updatedWidth - MIN_FRAME_SIZE ) /
+						( 1300 - MIN_FRAME_SIZE )
+				)
+			);
+
+		// Calculate the height based on the intermediate aspect ratio
+		// ensuring the frame arrives at a 9:19.5 aspect ratio.
+		const intermediateAspectRatio = lerp(
+			initialAspectRatioRef.current,
+			9 / 19.5,
+			lerpFactor
+		);
+		const newHeight = updatedWidth / intermediateAspectRatio;
+
+		setIsResizing( true );
+
+		if ( event.clientX < 344 ) {
+			const oversizeWidth = Math.max(
+				initialComputedWidthRef.current +
+					( updatedWidth - initialComputedWidthRef.current ) / 2,
+				initialComputedWidthRef.current
+			);
+
+			setFrameSize( {
+				width: oversizeWidth,
+				height: '100%',
+			} );
+			setIsOversized( true );
+		} else {
+			setFrameSize( { width: updatedWidth, height: newHeight } );
+			setIsOversized( false );
+		}
+	};
+
+	// Make the frame full screen when the user resizes it to the left.
+	const handleResizeStop = ( event ) => {
+		// Reset the initial aspect ratio if the frame is resized slightly
+		// above the sidebar but not far enough to trigger full screen.
+		if ( event.clientX < 344 && event.clientX > 200 ) {
+			setFrameSize( { width: '100%', height: '100%' } );
+		}
+
+		// Trigger full screen if the frame is resized far enough to the left.
+		if ( event.clientX < 200 ) {
+			setCanvasMode( 'edit' );
+			// Wait for the frame to animate to full screen before resetting its size.
+			const timeoutId = setTimeout( () => {
+				setFrameSize( { width: '100%', height: '100%' } );
+			}, 500 );
+
+			// Clean up the timeout when the effect is no longer needed.
+			return () => clearTimeout( timeoutId );
+		}
+
+		setIsResizing( false );
+	};
+
+	return (
+		<ResizableBox
+			as={ motion.div }
```

Comment:
> I would try to avoid mixing `motion` and `ResizableBox` via polymorphism (`as` prop).

I realized that, in the current implementation, it is unfortunately _necessary_ that the ResizableBox is also a `motion.div` 😬 In `re-resizable`, controlled `size` prop changes aren't reflected at all [until](https://github.com/bokuweb/re-resizable/blob/da458ab823949ee1090316a3d50950ff89a04a7d/src/index.tsx#L872-L874) `resizeStop`, meaning that we cannot do controlled size changes _during_ a resize gesture. Our controlled height changes are actually being executed through the `motion.div`'s `height`, not `re-resizable`'s `size` 😓

There is a possibility we could move the height management to `re-resizable` by giving it dynamic `lockAspectRatio` values (similar to what we do with `resizeRatio`). But since it is functional right now, I recommend deferring that to a nice-to-have follow-up.